File: Week 1/python/7.taking_input_all.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: Reading All Common Input Types
File: 7.taking_input_all.py

CONCEPT:
Java's Scanner offers `nextInt`, `nextLong`, `nextDouble`, `next`, `nextLine`.
Python uses `input()` plus type conversions. Below we demonstrate the
equivalent helpers for every common Java Scanner method.

KEY POINTS:
- `int(input())`         -> Java's nextInt()
- `int(input())`         -> Java's nextLong()  (Python int has no fixed size)
- `float(input())`       -> Java's nextFloat() / nextDouble()
- `input().split()[0]`   -> Java's next() (first whitespace-separated token)
- `input()`              -> Java's nextLine()
- `input()[0]`           -> Java has no nextChar(); idiom: `s.next().charAt(0)`
- For many tokens at once, use `sys.stdin.read().split()`

SYNTAX:
i  = int(input())
lo = int(input())                # arbitrary precision — replaces nextLong
f  = float(input())              # replaces nextFloat / nextDouble
tok = input().split()[0]         # first token
line = input()                   # whole line
ch  = input()[0]                 # first character of the line
"""

import sys
from typing import Iterator


def token_stream() -> Iterator[str]:
    """Yield whitespace-separated tokens from stdin lazily."""
    for line in sys.stdin:
        for tok in line.split():
            yield tok


def demo_with_tokens() -> None:
    """Use a token stream similar to how Java's Scanner returns tokens."""
    tokens = token_stream()

    a = int(next(tokens))         # nextInt
    b = int(next(tokens))         # nextInt
    c = a + b
    print(f"a + b = {c}")

    word = next(tokens)            # next() -> single token
    ch = word[0]                   # first character
    print(f"First char of '{word}': '{ch}'")

    d = float(next(tokens))        # nextDouble
    print(f"Double = {d}")

    big = int(next(tokens))        # nextLong / arbitrary-precision int
    print(f"Long  = {big}")


def demo_mock() -> None:
    """Same logic but on hard-coded data when no stdin is present."""
    raw = "10 5 hello 3.14 12345678901234567890"
    parts = raw.split()
    print(f"(mock input) {raw}")
    a, b = int(parts[0]), int(parts[1])
    print(f"a + b = {a + b}")
    word = parts[2]
    print(f"First char of '{word}': '{word[0]}'")
    print(f"Double = {float(parts[3])}")
    print(f"Long  = {int(parts[4])}")


def main() -> None:
    if sys.stdin.isatty():
        demo_mock()
    else:
        demo_with_tokens()


if __name__ == "__main__":
    main()


# NOTES:
# - Java distinguishes int vs long; Python has ONE int type with arbitrary precision.
# - Java distinguishes float vs double; Python has ONE float type (64-bit IEEE 754).
# - Python has NO Scanner pitfall around `nextLine()` after `nextInt()` because
#   conversions are done explicitly.
# - For competitive-programming-grade input, replace `input()` with
#   `sys.stdin.readline()` and read all tokens up-front.

File: Week 1/cpp/7.taking_input_all.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: Reading All Common Input Types
 * File: 7.taking_input_all.cpp
 *
 * CONCEPT:
 * Java's `Scanner` exposes one method per type (nextInt, nextDouble, etc.).
 * C++ relies on operator `>>` to pick the right parse based on the
 * destination variable's type. The end result is the same.
 *
 * KEY POINTS:
 *  - `>>` is overloaded for char, int, long, long long, float, double, string, ...
 *  - `cin >> w`   where w is a std::string reads ONE whitespace-delimited token.
 *  - `getline(cin, s)` reads a full line.
 *  - For a "single char" use `cin >> ch` (skips whitespace) or `cin.get(ch)`.
 *  - Always check that the extraction succeeded: `if (!(cin >> x)) ...`.
 *
 * SYNTAX:
 *   int i;          cin >> i;
 *   long long lo;   cin >> lo;
 *   double  d;      cin >> d;
 *   std::string w;  cin >> w;          // one token
 *   std::getline(cin, line);            // one line
 *
 * DRY RUN:
 *  Stdin: "10 5 hello 3.14 12345678901234567890"
 *    a=10, b=5, c=15
 *    word="hello", ch='h'
 *    d=3.14, lo=very big number
 */

#include <iostream>
#include <string>
#include <limits>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int a, b;
    std::cin >> a >> b;            // nextInt, nextInt
    int c = a + b;
    std::cout << "a + b = " << c << '\n';

    std::string word;
    std::cin >> word;              // next() — one token
    char ch = word.empty() ? '?' : word[0];
    std::cout << "First char of '" << word << "' = '" << ch << "'\n";

    double d;
    std::cin >> d;                 // nextDouble
    std::cout << "double = " << d << '\n';

    long long lo;
    std::cin >> lo;                // nextLong
    std::cout << "long long = " << lo << '\n';

    return 0;
}

/*
 * NOTES:
 *  - C++ requires <string> for std::string and <limits> for buffer draining.
 *  - On failed parse, `cin` enters a fail state and subsequent reads silently fail
 *    until you call `cin.clear()` and discard the bad input.
 *  - For very fast input, disable C-stdio sync (shown above) — this is the C++
 *    equivalent of Python's `sys.stdin.readline()` trick.
 *  - C++ has fixed-width long long (64 bits); Python int / Java BigInteger are
 *    arbitrary precision.
 */
